SharedBufferBuilder::take() should return a contiguous SharedBuffer if it only contains a single DataSegment.

jyavenard opened the pull request:
#### a20b10db9c2de64a37e2eb32579ec465db2bb73d
<pre>
SharedBufferBuilder::take() should return a contiguous SharedBuffer if it only contains a single DataSegment.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296684">https://bugs.webkit.org/show_bug.cgi?id=296684</a>
<a href="https://rdar.apple.com/157093750">rdar://157093750</a>

Reviewed by Youenn Fablet.

Make SharedBufferBuilder always return a SharedBuffer if the data it holds
is contiguous, and a FragmentedSharedBuffer otherwise.
Following this change, you can no longer create an object of the base class
FragmentedSharedBuffer if it is contiguous: it will always be a SharedBuffer
instead.
A side benefits is that it allows to remove a memory allocation whenever
we want to convert a FragmentedSharedBuffer into a SharedBuffer such as when
when using a SharedBufferBuilder and appending a single segment to it:
SharedBufferBuilder::takeAsContiguous() is now a no-op if the underlying
buffer is already contiguous.

API tests added.

* Source/WTF/wtf/TypeTraits.h: Adding RemoveCVSmartPointer convenience template.
* Source/WebCore/Modules/highlight/AppHighlight.h:
(IPC::AsyncReplyError&lt;WebCore::AppHighlight&gt;::create): Remove use of FragmentedSharedBuffer
that was in effect always contiguous.
* Source/WebCore/Modules/highlight/AppHighlightRangeData.cpp:
(WebCore::AppHighlightRangeData::create): Ditto
(WebCore::AppHighlightRangeData::toSharedBuffer const):
* Source/WebCore/Modules/highlight/AppHighlightRangeData.h:
* Source/WebCore/Modules/highlight/AppHighlightStorage.cpp: Ditto
(WebCore::AppHighlightStorage::restoreAndScrollToAppHighlight):
* Source/WebCore/Modules/highlight/AppHighlightStorage.h: Ditto

Canonical link: <a href="https://commits.webkit.org/298168@main">https://commits.webkit.org/298168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d80e03f2121fb53ae4029d947da0727ef2bce7dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34233 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24696 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120653 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34862 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87010 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117435 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/27776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67401 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26960 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20943 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64333 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106883 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21057 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123858 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113050 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41500 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30968 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99007 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95613 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24361 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40775 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18612 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37532 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41379 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46889 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137267 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40967 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36705 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44274 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42717 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -528,21 +474,46 @@ SharedBuffer::SharedBuffer(Ref<const DataSegment>&& segment)
+{
+    m_size = fileData.size();
+    m_segments.append({ 0, DataSegment::create(WTFMove(fileData)) });
+    m_contiguous = true;
```
Can we add a protected FragmentSharedBuffer constructor to allow initialising the members andrewrite it to something like:
```explicit FragmentSharedBuffer(Ref<DataSegment>&& segment)
    : m_size(segment->size()
    , m_segments(DataSegmentVector::from({ 0, WTFMove(segment) })
    , m_contiguous(true)````

jyavenard replied:
agree. indeed simplifies the code a fair bit too.

youennf commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -241,33 +220,16 @@ class FragmentedSharedBuffer : public ThreadSafeRefCounted<FragmentedSharedBuffe
 
+    size_t m_size { 0 };
     DataSegmentVector m_segments;
     bool m_contiguous { false };
```
We do not really like protected members as per C++ guide, can we do without it?

jyavenard replied:
this is existing and has been there for a while. how would you want them to be instead?

youennf commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -669,11 +638,13 @@ void SharedBufferBuilder::initialize(Ref<FragmentedSharedBuffer>&& buffer)
         return;
     }
     append(buffer);
+    if (m_buffer->segmentsCount() <= 1)
```
I would tend to rewrite it by checking upfront isContiguous and either moving or cloning.
Otherwise, `m_buffer->m_contiguous = m_buffer->segmentsCount() <= 1`
Also, it seems `SharedBufferBuilder::append` is already setting m_contiguous.

jyavenard replied:
yes, but this can be called from the SharedBuffer::create , independently from the SharedBufferBuilder

youennf commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -357,7 +340,10 @@ class SharedBufferBuilder {
-        : m_buffer(FragmentedSharedBuffer::create(std::forward<Args>(args)...)) { }
+        : m_buffer(SharedBuffer::create(std::forward<Args>(args)...))
+    {
+        m_buffer->m_contiguous = m_buffer->segmentsCount() <= 1;
```
It would be nice if `SharedBuffer::create` would set its m_contiguous value itself.

youennf commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -367,6 +353,7 @@ class SharedBufferBuilder {
     {
         ensureBuffer();
         Ref { *m_buffer }->append(std::forward<Args>(args)...);
+        m_buffer->m_contiguous = m_buffer->segmentsCount() <= 1;
```
Ditto here, it seems like m_buffer would set m_contiguous itself and not the builder.

jyavenard replied:
append is a FragmentedSharedBuffer method ; it doesn't know if it's called from SharedBuffer or not. a FragmentedSharedBuffer that is not actually a SharedBuffer, should never have m_contiguous set to true.

The alternative would be to make isContiguous() a virtual method instead and do without m_contiguous

youennf commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -125,36 +95,12 @@ std::optional<Ref<FragmentedSharedBuffer>> FragmentedSharedBuffer::fromIPCData(I
+    : m_size(segment->size())
 {
-    return adoptRef(*new FragmentedSharedBuffer(mappedBuffer));
+    m_segments.append({ 0, WTFMove(segment) });
```
Can we init m_segments with segment using Vector::from()?

youennf commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -559,15 +512,17 @@ RefPtr<SharedBuffer> SharedBuffer::createWithContentsOfFile(const String& filePa
+    return segments()[0].segment->span();
 }
 
 const uint8_t& SharedBuffer::operator[](size_t i) const
```
```suggestion
uint8_t SharedBuffer::operator[](size_t i) const
```
?

youennf commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -559,15 +512,17 @@ RefPtr<SharedBuffer> SharedBuffer::createWithContentsOfFile(const String& filePa
-    RELEASE_ASSERT(i < size() && !m_segments.isEmpty());
-    return m_segments[0].segment->span()[i];
+    ASSERT(isContiguous());
+    RELEASE_ASSERT(i < size() && segmentsCount());
```
Is this still necessary?
Won't we crash anyway if doing segments()[0] and segments is empty (and ditto for accessing outside of span())?

youennf commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -334,12 +303,29 @@ class SharedBuffer : public FragmentedSharedBuffer {
+    friend class SharedBufferBuilder;
+
+    SharedBuffer() { }
     SharedBuffer(const DataSegment& segment)
```
explicit.

youennf commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -292,24 +256,29 @@ class FragmentedSharedBuffer : public ThreadSafeRefCounted<FragmentedSharedBuffe
+    WEBCORE_EXPORT static Ref<SharedBuffer> create(GstMappedOwnedBuffer&);
+#endif
+#if USE(SKIA)
+    WEBCORE_EXPORT static Ref<SharedBuffer> create(sk_sp<SkData>&&);
```
Maybe many of these create methods could be replaced by a templated version that would call `DataSegment::create()` and then call SharedBuffer::Create(Ref<const DataSegment>&&)`

jyavenard replied:
I'll take this as a follow-up

youennf reviewed (approved):
This is an improvement.
As discussed offline, I would tend to go with 3 classes:
- MutableFragmentedSharedBuffer
- FragmentedSharedBuffer (immutable)
- SharedBuffer (immutable, contiguous)

Maybe MutableFragmentedSharedBuffer could become just a Vector instead.
And we could break the inheritance of SharedBuffer and FragmentedSharedBuffer.
And SharedBufferBuilder would only create immutable buffers.

jyavenard committed:
SharedBufferBuilder::take() should return a contiguous SharedBuffer if it only contains a single DataSegment.
https://bugs.webkit.org/show_bug.cgi?id=296684
rdar://157093750

Reviewed by Youenn Fablet.

Make SharedBufferBuilder always return a SharedBuffer if the data it holds
is contiguous, and a FragmentedSharedBuffer otherwise.
Following this change, you can no longer create an object of the base class
FragmentedSharedBuffer if it is contiguous: it will always be a SharedBuffer
instead.
A side benefits is that it allows to remove a memory allocation whenever
we want to convert a FragmentedSharedBuffer into a SharedBuffer such as when
when using a SharedBufferBuilder and appending a single segment to it:
SharedBufferBuilder::takeAsContiguous() is now a no-op if the underlying
buffer is already contiguous.

API tests added.

* Source/WTF/wtf/TypeTraits.h: Adding RemoveCVSmartPointer convenience template.
* Source/WebCore/Modules/highlight/AppHighlight.h:
(IPC::AsyncReplyError<WebCore::AppHighlight>::create): Remove use of FragmentedSharedBuffer
that was in effect always contiguous.
* Source/WebCore/Modules/highlight/AppHighlightRangeData.cpp:
(WebCore::AppHighlightRangeData::create): Ditto
(WebCore::AppHighlightRangeData::toSharedBuffer const):
* Source/WebCore/Modules/highlight/AppHighlightRangeData.h:
* Source/WebCore/Modules/highlight/AppHighlightStorage.cpp: Ditto
(WebCore::AppHighlightStorage::restoreAndScrollToAppHighlight):
* Source/WebCore/Modules/highlight/AppHighlightStorage.h: Ditto

Canonical link: https://commits.webkit.org/298168@main
